### DISCUSSION TITLE
Terminal welcome message

### DISCUSSION THREAD

**Discussion by ptyork** [community/community#43534] [Codespaces] [answered]
### Select Topic Area

Question

### Body

When you first open a custom codespace, it prints a few lines of welcome that start something text like:

`Welcome to Codespaces! You are on a custom image defined in your devcontainer.json file.`

Is there a way to customize this for a custom template? I'm creating spaces for use in GitHub classrooms for a less technically savvy set of students, and I'd prefer to present a more appropriate message here if possible.

**Comment by jfullstackdev** [comment]
this is the official page 

https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/introduction-to-dev-containers

**Comment by auchynnikau** [comment]
To change the default welcome message in the console of GitHub CodeSpaces environment, you can create a file named `.welcome` in the root of your repository. You can include any text or commands that you want. When you save the file and push it to your repository, the new welcome message will be displayed the next time you open the environment.

It will also work for a CodeSpaces environment created from a custom template that includes `.welcome` file.

**Reply by ptyork** [reply]
Thanks. Unfortunately, that doesn't seem to work:
![Untitled](https://user-images.githubusercontent.com/1051795/211167417-f00433e4-c49b-4b62-b814-00bad32c7aa2.png)
Tried it both in the root and in .devcontainer/. Tried without and then with the file actually committed. Tried a prebuild. None seemed to pick up that .welcome file and replace the default message.

**Reply by auchynnikau** [reply]
You can try another approach. Open the `devcontainer.json` and find the `postCreateCommand` property and replace the default command with this one:

```
"postCreateCommand": "echo 'Your custom message'"
```

**Reply by RafaelNevesdeOliveira** [reply]
nice

**Comment by aditya30394** [comment]
Hey @ptyork, Thanks for writing to us! 

Unfortunately, that message is fixed as of now and there is no plan to provide a way to customize this one in the near future.

**Reply by ptyork** [reply]
Thank you for the definitive answer. I guess as feedback, as this is now integrated with GitHub Classroom, it'd be nice to have this extra bit of flexibility...minimally to be able to disable it with a flag in devcontainer.json. The messages make sense for corporate users, but just add confusion in classroom assignments. Thanks!

**Comment by ptyork** [comment, marked as answer]
Okay, with enough perseverance (and snooping), one can come up with (i.e., "hack") a solution. Two, actually. First option:
```
	"postCreateCommand": "mkdir -p $HOME/.config/vscode-dev-containers/ && touch $HOME/.config/vscode-dev-containers/first-run-notice-already-displayed"
```
This disables the "first-run" special message.

Second option:
```
	"postCreateCommand": "sudo echo 'ENTER A SPECIAL MESSAGE HERE' > /workspaces/.codespaces/shared/first-run-notice.txt"
```
This replaces the "first-run" special message. Could obviously use a `sudo cp` in place of the `sudo echo` if you wanted to have a more complex first run notice.

Not a fan of hacks, but at least it does the trick until there's a supported solution.

**Reply by JustinGrote** [reply]
Hmm, didn't work for me in Github Codespaces. This was after a full rebuild.
![image](https://user-images.githubusercontent.com/15258962/215181186-0280483d-dcb0-4dba-b39a-41f58e4eb0a1.png)

**Reply by ptyork** [reply]
Hmm. It's possible that a different container base image works differently, I suppose. Check your /workspaces/.codespaces/shared/first-run-notice.txt file. If it exists and looks like the welcome message, perhaps try the second option. Just create an empty message instead:

	"postCreateCommand": "sudo echo > /workspaces/.codespaces/shared/first-run-notice.txt"

**Reply by cr2007** [reply]
The 2nd step is not working either.

I added the following in the `postCreateCommand` section within the `.devcontainer.json` file.

```json
"postCreateCommand": "sudo cp .devcontainer/welcome.txt /workspaces/.codespaces/shared/first-run-notice.txt"
```

And it didn't change much, it still produced the generic message.

I tried it even after a full rebuild like how @JustinGrote did too.

**Reply by ptyork** [reply]
All of the base distros appear to add the common-utils feature. This is the source of the offending message. It runs this script:

https://github.com/devcontainers/features/blob/main/src/common-utils/scripts/rc_snippet.sh

If you look at the order of operations in there, it appears that text in `/usr/local/etc/vscode-dev-containers/first-run-notice.txt` takes precedence over text in `/workspaces/.codespaces/shared/first-run-notice.txt`. For my Ubuntu base image, the latter works fine. Though you might try directly overwriting the `/usr/local...` version instead.

**Reply by cr2007** [reply]
What about using `onCreateCommand` instead of `postCreateCommand` ?

As mentioned in the [devcontainer.json docs](https://containers.dev/implementors/json_reference/#_lifecycle-scripts:~:text=vs%20object%20properties.-,onCreateCommand%20%F0%9F%8F%B7%EF%B8%8F,-string%2C%0Aarray):

> This command is the first of three (along with updateContentCommand and postCreateCommand) that finalizes container setup when a dev container is created. It and subsequent commands execute inside the container immediately after it has started for the first time.
> 
> Cloud services can use this command when caching or prebuilding a container. This means that it will not typically have access to user-scoped assets or secrets.
> 
> Note that the array syntax will execute the command without a shell. You can [learn more](https://containers.dev/implementors/json_reference/#formatting-string-vs-array-properties) about formatting string vs array vs object properties.

**Reply by cr2007** [reply]
@ptyork 

There was no `first-run-notice.txt` file in `/workspaces/.codespaces/shared`, so the former worked for me.

The following worked just right in the `devcontainer.json` file:

```json
"postCreateCommand": "sudo cp .devcontainer/welcome.txt /usr/local/etc/vscode-dev-containers/first-run-notice.txt"
```

If this helps for anyone, just create a `welcome.txt` file containing the message you want to put in the terminal.
And put the above line in the `devcontainer.json` file.

And it should work like a charm.

**Comment by sina1100** [comment]
Hello, I received a serial email from this address and I am suspicious, if it is possible, check it

**Comment by cr2007** [comment]
The below thing worked just fine in my `devcontainer.json` file.

```json
"postCreateCommand": "sudo cp .devcontainer/welcome.txt /usr/local/etc/vscode-dev-containers/first-run-notice.txt"
```

Basically you have to create a `welcome.txt` file containing the message you want to put in the terminal.
And put the above line in the devcontainer.json file.

And it should work like a charm.

<img width="428" alt="image" src="https://github.com/community/community/assets/73425927/d7f74362-3cea-4e36-9653-17dfd66342fe">

**Comment by berkay-tec** [comment]
Thanks for the helpful comments. 

As of July 2025, `/usr/local/etc/vscode-dev-containers/first-run-notice.txt` **works only** for **dev containers**. I needed to make it run for Codespaces. 

Solution: to use `/workspaces/.codespaces/shared/first-run-notice.txt `

1. I needed to generate **dynamic** content, so I created a shell script named `generate_motd.sh` with content:
```bash
#!/bin/sh
python -m misc.get_dynamic_welcome_message | sudo tee /workspaces/.codespaces/shared/first-run-notice.txt > /dev/null
```

and in `devcontainer.json`
```json
"postCreateCommand": "sh .devcontainer/generate_motd.sh",
```
